Add instrument allocations pass

loganek opened the pull request:
This pass wraps `malloc`/`calloc`/`free`/`realloc` around `trace_{func_name}` functions so we can keep track on all the allocations. I understand this pass won't work for some of the languages, but opening a PR in case someone finds it useful and worth enough including in wasm-opt. 
Also, I don't have a ton of experience writing compiler passes, so any suggestion is very appreciated. 

Related issue: https://github.com/WebAssembly/binaryen/issues/6548

kripken commented on src/passes/InstrumentAllocations.cpp:
```diff
@@ -0,0 +1,158 @@
+     {Type::i32, Type::i32, Type::i32},
+     Type::none},
+    {Name("free"), Name("trace_free"), {Type::i32}, Type::none}};
+};
```
Perhaps this could be an option? I am thinking of something like
```
wasm-opt --trace-calls=malloc,i32,i32;calloc,i32,i32,i32;realloc,i32,i32,i32;free,none,i32
```
So the user provides a list of which functions to trace and their types. That would make this pretty general and maybe there are more use cases?

(For an example of a pass that gets an argument see `ExtractFunction`).

loganek replied:
I like the idea of having the pass more generic. I've updated the code according to your suggestion. Since the tracer receives both return value and parameters as function arguments, the pass should never expects to receive `none`; instead it expects just a sequence of (optional) return value and (optional) list of parameters.

kripken commented on test/passes/instrument-allocations.txt:
```diff
@@ -0,0 +1,82 @@
+   (local.get $1)
+  )
+ )
+)
```
Another option rather than instrumenting around each function call is to add new wrapper functions. That would be better for code size (existing calls remain the same size, they just point to the wrappers, and the wrappers are just a fixed increase in size). The inliner can inline them if the user then optimizes (and the inliner thinks it is worth it), which could make it just as fast.

But if size/speed are not important here then the current approach may be fine.

loganek replied:
I personally only intend to use it for debugging and not in the production code (at at least so far) so neither speed nor size is critical. I think the pass can be updated once there are different usecases.

kripken commented on src/passes/TraceCalls.cpp:
```diff
@@ -0,0 +1,229 @@
+// Instruments the build with code to intercept selected function calls.
+// This can be e.g. used to trace allocations (malloc, free, calloc, realloc)
+// and build tools for memory usage analysis.
+// The pass supports SIMD but multi-value feature is not supported yet.
```
```suggestion

// The pass supports SIMD but the multi-value feature is not supported yet.
```

kripken commented on src/passes/TraceCalls.cpp:
```diff
@@ -0,0 +1,229 @@
+  }
+};
+
+// declare passes
+
```
```suggestion
```

kripken replied:
Looks like we have this comment in other passes when there are multiple passes in a file, but not when there is just one, like here.

kripken commented on test/passes/trace-calls.wast:
```diff
@@ -0,0 +1,45 @@
+  (func $test_dont_trace_me
+    call $donttraceme
+  )
+)
```
We are moving tests from `passes/` to `lit/passes/`, and converting them to lit, so new passes should go there, sorry for the confusion.

If you're not familiar with LLVM lit then copy-pasting from other files in that directory should be easy enough.

loganek replied:
no problem, done.

kripken commented on src/passes/TraceCalls.cpp:
```diff
@@ -0,0 +1,229 @@
+  void visitCall(Call* curr) {
+    auto* target = getModule()->getFunction(curr->target);
+
+    auto function_it = std::find_if(tracedFunctions.begin(),
```
`function_it` etc. should be `functionIt` in terms of naming conventions. However, in this case it can be just `iter` which is what we tend to use for a quick iterator.

kripken commented on src/passes/TraceCalls.cpp:
```diff
@@ -0,0 +1,228 @@
+}
+
+static Type strToType(const std::string& strType) {
+  if (strType.compare("i32") == 0)
```
```suggestion
  if (strType == "i32") {
```
(`==` is more concise, and also we always require `{ }` around if bodies.

kripken commented on src/passes/TraceCalls.cpp:
```diff
@@ -0,0 +1,228 @@
+    for (const auto& definition : String::Split(arg, ";")) {
+      auto parts = String::Split(definition, ",");
+      if (parts.size() == 0) {
+        // empty definition, ignore
```
```suggestion
        // Empty definition, ignore.
```

kripken commented on test/lit/passes/trace-calls.wast:
```diff
@@ -0,0 +1,124 @@
+;; RUN: foreach %s %t wasm-opt --enable-simd --trace-calls="noparamsnoresults;singleparamnoresults,f64;multiparamsnoresults:tracempnr,i32,i64,f32;noparamssingleresult,v128;multiparamssingleresult,v128,i32,v128" -S -o - | filecheck %s
```
There should be a line here like this:
```
;; NOTE: Assertions have been generated by update_lit_checks.py --all-items and should not be edited.
```
That allows the auto-update script to handle it. After adding that run `./scripts/update_lit_checks.py test/lit/passes/trace-calls.wast`

kripken reviewed (commented):
Is there an explanation of the commandline format somewhere? If it's short enough in the `--help` that would be good, but I think it might not be, so the `.cpp` file perhaps.

However, a second question: Why does the commandline format have the types in there? It seems like we can detect the types from the function name? But I may have forgotten something here.

loganek commented:
> Is there an explanation of the commandline format somewhere? If it's short enough in the `--help` that would be good, but I think it might not be, so the `.cpp` file perhaps.

With the latest update where I removed the need for specifying the types, I think it's simple enough; there was already a help string defined in the TraceCalls.cpp in the `getArgument()` - it's not printed on `--help`, but it's displayed when the user tries to incorrectly use the pass. Do you think this should also be displayed when user calls `--help`? I'm not sure I saw it for other passes (e.g. `extract-function-index`).

> However, a second question: Why does the commandline format have the types in there? It seems like we can detect the types from the function name? But I may have forgotten something here.

That's an excelent point; so initially the tracer was for pre-defined functions (malloc,calloc, realloc, free) but I changed that after the suggestion in the comment https://github.com/WebAssembly/binaryen/pull/6619#discussion_r1610631527 to make the pass more generic; I also blindly followed the recommendation to include types as parameters but it actually is not needed, and we can infer the type from the function being traced. I updated the PR now, thanks.

kripken commented on test/lit/passes/trace-calls.wast:
```diff
@@ -0,0 +1,151 @@
+  ;; CHECK-NEXT:  (call $donttraceme)
+  ;; CHECK-NEXT: )
+  (func $test_dont_trace_me
+    call $donttraceme
```
```suggestion
    (call $donttraceme)
```
(just to be consistent, it's valid either way)

kripken commented on test/lit/passes/trace-calls.wast:
```diff
@@ -0,0 +1,151 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py --all-items and should not be edited.
+
+;; RUN: wasm-opt --enable-simd --trace-calls="noparamsnoresults,singleparamnoresults,multiparamsnoresults:tracempnr,noparamssingleresult,multiparamssingleresult" %s -S -o - | filecheck %s
```
I don't seem to see an explanation for this notation yet? I suggested before either putting that in the cpp file or the `--help` output.

I think an explanation is needed because I can guess that `,` separates functions, but it's not obvious what `:` does...

loganek replied:
I've added explanation in the CPP file, I hope that's enough but let me know if you'd like this to be included in the help string as well

kripken reviewed (commented):
Looks much better!

kripken commented on src/passes/TraceCalls.cpp:
```diff
@@ -0,0 +1,218 @@
+//   (local $1 i32)
+//   (local $2 i32)
+//   (block (result i32)
+//     (call $trace_allocation
```
```suggestion
//     (call $trace_alloc
```
To match the comment text, unless I've misunderstood?

loganek replied:
yep, that was my miss, thanks.

kripken commented:
Looks like there are some CI errors that need to be resolved before landing.

loganek commented:
@kripken the build is fixed now. Tests were failing because I iterated through `unordered_map` to create imports which is non-deterministic; I switched to `map` which doesn't make significant difference in this specific case, but I'm curious how do you normally deal with this type of nondeterminism when you write LIT tests?

kripken commented:
@loganek Iterating on a map here makes sense, yeah, as the keys are Names. In general we sometimes fix nondeterminism that way, if there is a natural key to use. If there aren't, say if the map is from `Expression*`, then we need something else (since pointers change in every execution), and then we have to find another way. One mechanism is [InsertOrderedSet](https://github.com/WebAssembly/binaryen/blob/main/src/support/insert_ordered.h) which uses the insertion order to be deterministic (assuming that order in fact is).
